patches/server/0918-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,826 @@
+-        return CraftEventFactory.handleNonLivingEntityDamageEvent(entity, source, damage, true);
++    public static boolean handleNonLivingEntityDamageEvent(Entity entity, DamageSource source, io.papermc.paper.entity.damage.DamageOrigin origin, double damage) { // Paper - Damage source wrapper
++        return CraftEventFactory.handleNonLivingEntityDamageEvent(entity, source, origin, damage, true);
+     }
+ 
+-    public static boolean handleNonLivingEntityDamageEvent(Entity entity, DamageSource source, double damage, boolean cancelOnZeroDamage) {
+-        return CraftEventFactory.handleNonLivingEntityDamageEvent(entity, source, damage, cancelOnZeroDamage, false);
++    public static boolean handleNonLivingEntityDamageEvent(Entity entity, DamageSource source, io.papermc.paper.entity.damage.DamageOrigin origin, double damage, boolean cancelOnZeroDamage) { // Paper - Damage source wrapper
++        return CraftEventFactory.handleNonLivingEntityDamageEvent(entity, source, origin, damage, cancelOnZeroDamage, false);
+     }
+ 
+-    public static boolean handleNonLivingEntityDamageEvent(Entity entity, DamageSource source, double damage, boolean cancelOnZeroDamage, boolean cancelled) {
++    public static boolean handleNonLivingEntityDamageEvent(Entity entity, DamageSource source, io.papermc.paper.entity.damage.DamageOrigin origin, double damage, boolean cancelOnZeroDamage, boolean cancelled) { // Paper - Damage source wrapper
+         final EnumMap<DamageModifier, Double> modifiers = new EnumMap<DamageModifier, Double>(DamageModifier.class);
+         final EnumMap<DamageModifier, Function<? super Double, Double>> functions = new EnumMap(DamageModifier.class);
+ 
+         modifiers.put(DamageModifier.BASE, damage);
+         functions.put(DamageModifier.BASE, ZERO);
+ 
+-        final EntityDamageEvent event = CraftEventFactory.handleEntityDamageEvent(entity, source, modifiers, functions, cancelled);
++        final EntityDamageEvent event = CraftEventFactory.handleEntityDamageEvent(entity, source, origin, modifiers, functions, cancelled); // Paper - Damage source wrapper
+ 
+         if (event == null) {
+             return false;
+diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+index 7c1e510a0ad4b69c1fedc3367a8216046efb228e..1eab4ab2c9312eb61195da62ef1f63765d2c5138 100644
+--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
++++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+@@ -604,6 +604,11 @@ public final class CraftMagicNumbers implements UnsafeValues {
+         Preconditions.checkArgument(material.isBlock(), material + " is not a block");
+         return getBlock(material).hasCollision;
+     }
++
++    @Override
++    public io.papermc.paper.entity.damage.DamageOrigin getDamageOrigin(String key) {
++        return net.minecraft.world.damagesource.DamageSource.staticDamageBinder.get(key);
```

Comment:
Store this in a separate class, don't wanna put diffs in the nms classes like this if we don't have to.